clippy_lints/src/vec.rs

```diff
@@ -110,31 +97,62 @@ impl<'tcx> LateLintPass<'tcx> for UselessVec {
             .is_continue();
 
             if only_slice_uses {
-                self.check_vec_macro(
-                    cx,
-                    &vec_args,
-                    expr.span.ctxt().outer_expn_data().call_site,
-                    SuggestedType::Array,
-                );
+                let span = expr.span.ctxt().outer_expn_data().call_site;
+                self.lintable_exprs.insert(expr.hir_id);
+                self.check_vec_macro(cx, &vec_args, span, expr.hir_id, SuggestedType::Array);
             }
         }
-
         // search for `for _ in vec![…]`
-        if let Some(higher::ForLoop { arg, .. }) = higher::ForLoop::hir(expr)
+        else if let Some(higher::ForLoop { arg, .. }) = higher::ForLoop::hir(expr)
             && let Some(vec_args) = higher::VecArgs::hir(cx, arg)
             && self.msrv.meets(msrvs::ARRAY_INTO_ITERATOR)
         {
             // report the error around the `vec!` not inside `<std macros>:`
             let span = arg.span.ctxt().outer_expn_data().call_site;
-            self.check_vec_macro(cx, &vec_args, span, SuggestedType::Array);
+            self.lintable_exprs.insert(arg.hir_id);
+            self.check_vec_macro(cx, &vec_args, span, arg.hir_id, SuggestedType::Array);
+        }
+        // search for `&vec![_]` or `vec![_]` expressions where the adjusted type is `&[_]`
+        else if let Some(vec_args) = higher::VecArgs::hir(cx, expr.peel_borrows()) {
+            let (suggest_slice, span) = if let ExprKind::AddrOf(BorrowKind::Ref, mutability, _) = expr.kind {
+                // `expr` is `&vec![_]`, so suggest `&[_]` (or `&mut[_]` resp.)
+                (SuggestedType::SliceRef(mutability), expr.span)
+            } else {
+                // `expr` is the `vec![_]` expansion, so suggest `[_]`
+                // and also use the span of the actual `vec![_]` expression
+                (SuggestedType::Array, expr.span.ctxt().outer_expn_data().call_site)
+            };
+
+            if adjusts_to_slice(cx, expr) {
+                self.check_vec_macro(cx, &vec_args, span, expr.hir_id, suggest_slice);
+            } else if !self.lintable_exprs.contains(&expr.hir_id) {
+                self.span_to_lint_map.insert(span, None);
+            }
+        }
+    }
+
+    fn check_crate_post(&mut self, cx: &LateContext<'tcx>) {
+        for (span, lint_opt) in &self.span_to_lint_map {
+            if let Some((hir_id, suggest_slice, snippet, applicability)) = lint_opt {
```

Comment:
The reason to insert `None` when we don't want to lint is to mark the specific `span` as non-lintable. There are cases where the same span expands to multiple expressions (when a macro expands the vec!) and we want to lint only when all of the expanded expressions are ok for linting.

The reason for `lintable_exprs` is because the first two checks of the lint will check cases where we do want to emit the lint even when the `vec![...]` expression doesn't adjust to a slice. I chose to keep track of those expressions separately, although I could have also added a parameter to `check_vec_macro` indicating whether we want to lint immediately, or wait until the `check_crate_post` stage to emit lints.

Reply:
There's no need to gather all lintable expressions in a single collection and then do a crate pass, you can just lint in place.

The infrastructure needed to collect all lintable expressions costs performance and memory, some things that we're trying to improve right now. If you're trying to improve code cleanliness, you can just refactor the `span_lint...` function into it's own function that takes less arguments. Like a `report` function or `emit`. It's done in other lints to avoid typing the same description / notes multiple times.

Reply:
Thanks for the suggestion. I'll look into refactoring this weekend, since I am busy with school this week.